filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobPathBuilder.scala

```diff
@@ -57,35 +59,71 @@ object BlobPathBuilder {
 
 class BlobPathBuilder(blobTokenGenerator: BlobTokenGenerator, container: String, endpoint: String) extends PathBuilder {
 
-  val credential: AzureSasCredential = new AzureSasCredential(blobTokenGenerator.getAccessToken)
-  val fileSystemConfig: Map[String, Object] = Map((AzureFileSystem.AZURE_STORAGE_SAS_TOKEN_CREDENTIAL, credential),
-                                                  (AzureFileSystem.AZURE_STORAGE_FILE_STORES, container),
-                                                  (AzureFileSystem.AZURE_STORAGE_SKIP_INITIAL_CONTAINER_CHECK, java.lang.Boolean.TRUE))
+  def build(string: String): Try[BlobPath] = {
+    validateBlobPath(string, container, endpoint) match {
+      case ValidBlobPath(path) => Try(BlobPath(path, endpoint, container, blobTokenGenerator))
+      case UnparsableBlobPath(errorMessage: Throwable) => Failure(errorMessage)
+    }
+  }
+  override def name: String = "Azure Blob Storage"
+}
 
-  def retrieveFilesystem(uri: URI): Try[FileSystem] = {
-    Try(FileSystems.getFileSystem(uri)) recover {
-      // If no filesystem already exists, this will create a new connection, with the provided configs
-      case _: FileSystemNotFoundException => FileSystems.newFileSystem(uri, fileSystemConfig.asJava)
+object BlobPath {
+
+  def buildURI(endpoint: String) = new URI("azb://?endpoint=" + endpoint)
+
+  def buildConfigMap(credential: AzureSasCredential, container: String): Map[String, Object] = {
+    Map((AzureFileSystem.AZURE_STORAGE_SAS_TOKEN_CREDENTIAL, credential),
+      (AzureFileSystem.AZURE_STORAGE_FILE_STORES, container),
+      (AzureFileSystem.AZURE_STORAGE_SKIP_INITIAL_CONTAINER_CHECK, java.lang.Boolean.TRUE))
+  }
+
+  def findNioPath(path: String, endpoint: String, container: String, blobTokenGenerator: BlobTokenGenerator, attempted: Boolean = false): NioPath = (for {
```

Comment:
Ideally we'd have all this checking and retrying within `retrieveFilesystem`. I'm guessing you're not doing that because we don't actually know whether the filesystem doesn't work until we call `checkAccess`, for which we need to know the Blob file path. Let's discuss!

Reply:
Ay, there's the rub.